### DISCUSSION TITLE
How to forward remote IP address for http in Blazor environment?

### DISCUSSION THREAD

**Discussion by OnSive** [ThreeMammals/Ocelot#1748] [Q&A] [answered]
When I send a ws request to the _Blazor Server Side_ then I get the real remote IP address of the remote device, but when the request is a http request to the _Rest API_ then the remote IP is `::1`.

Since I cannot test the application on a server without considerable effort, I wonder if it could have to do with the debug environment.

## Expected Behavior
When connecting to the server with a remote machine to get the remote ip address on http and ws request

## Actual Behavior
The ip address of the remote machine is returned as `::1` on http and `82.x.x.x` on ws.

## Specifications
  - Version: 20.0.0
  - Platform: Windows 10
  - Backend: Microsoft.NET.Sdk.Worker .NET 7.0, Microsoft.AspNetCore

## Configs:
ocelot.json
```json
"Routes": [
  {
    "__ServiceName": "Rest API",
    "DownstreamPathTemplate": "/api/{everything}",
    "DownstreamScheme": "http",
    "DownstreamHostAndPorts": [
      {
        "Host": "localhost",
        "Port": 5000
      }
    ],
    "UpstreamPathTemplate": "/api/{everything}",
    "UpstreamHttpMethod": [ "Get", "Delete", "Patch", "Put", "Post" ],
    "SwaggerKey": "crud"
  },
  {
    "__ServiceName": "Blazor Server Side",
    "DownstreamPathTemplate": "/{catchAll}",
    "DownstreamScheme": "ws",
    "DownstreamHostAndPorts": [
      {
        "Host": "localhost",
        "Port": 5001
      }
    ],
    "UpstreamPathTemplate": "/{catchAll}",
    "UpstreamHttpMethod": [ "GET", "POST", "PUT", "DELETE", "OPTIONS" ]
  }
```

#### IpForwardHandler : DelegatingHandler
```csharp
public sealed class IpForwardHandler : DelegatingHandler
{
    private readonly IHttpContextAccessor _contextAccessor;
    private readonly ILogger<IpForwardHandler> _logger;

    public IpForwardHandler(IHttpContextAccessor contextAccessor, ILogger<IpForwardHandler> logger)
    {
        _contextAccessor = contextAccessor;
        _logger = logger;
    }

    protected override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        AddIpHeader(request, cancellationToken);
        return base.Send(request, cancellationToken);
    }

    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        AddIpHeader(request, cancellationToken);
        return base.SendAsync(request, cancellationToken);
    }

    private void AddIpHeader(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        const string key = "X-Forwarded-For";
        if (!request.Headers.Any(x => x.Key == key))
        {
            var ip = _contextAccessor.HttpContext.Connection.RemoteIpAddress.ToString();
            request.Headers.Add(key, ip);
            _logger.LogInformation("Added IP {ip}", ip);
        }
    }
}
```

## Additional question
What would be a feasible way to forward the remote ip address?
as the:
```json
"UpstreamHeaderTransform": {
  "X-Forwarded-For": "{RemoteIpAddress}"
}
```
Configuration does not seem to do anything. (at least not in the ASP.NET backend)

Edit:
- Added information that the test was run between two different machines.

**Comment by raman-m** [comment, marked as answer]
Well...

>When I send a **ws** request to the _Blazor Server Side_
>Get the remote ip address on http and **ws** request

Are you saying about WebSocket requests? [Websockets](../blob/develop/docs/features/websockets.rst) feature?

-----
> Remote ip address is ::1 on http and "real" on ws.

Yeah, you see such facts because of different protocols, and they behave differently. Don't you ask us, as Ocelot team, to fix that? 🤣 

----
>```json
>     "DownstreamHostAndPorts": [
>       {
>         "Host": "localhost", // come on!
>         "Port": 5000
>       }
>     ],
> ```

You use `localhost` and local environment PC to host the solution, and you wonder why do you have localhost addresses?
Come on!
What is your upstream settings in global section? I mean this:
```json
    "GlobalConfiguration": {
        "BaseUrl": "somehost:someport"
    }
```
I guess, BaseUrl is also `localhost`, right? 😉 

---
> #### IpForwardHandler : DelegatingHandler

Why do you write special delegating handler to get remote address IP? It seems you didn't read [the documentation](https://ocelot.readthedocs.io/en/latest/index.html)...

**Reply by raman-m** [reply]
Sorry?!..

**Reply by OnSive** [reply]
Didn't see the edit while writing the comment.
I'm using two different machines (the development machine and a remote machine) to test the system, port is opened and the servers are reachable.

The issue with the remote IP address appears when sending requests from the remote machine.
When I tried using the Headers Transformation (about what you wrote in the other answer) feature I didn't event get the headers in the ASP.NET backed.

That's why I tried writing a delegate handler as no other option seemed to work and I get still the same not working result.

Here is everything I tried with the config - this is the combination of all tests I did, no normally I don't have duplicating configuration:
```
{
  "GlobalConfiguration": {
    "BaseUrl": "http://localhost:80",
    "UpstreamHeaderTransform": {
      "Referrer": "{RemoteIpAddress}",
      "X-Forwarded-For": "{RemoteIpAddress}"
    }
  },
  "Routes": [
    {
      "__ServiceName": "Rest API Service",
      "DownstreamPathTemplate": "/api/{everything}",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 5000
        }
      ],
      "UpstreamPathTemplate": "/api/{everything}",
      "UpstreamHttpMethod": [ "Get", "Delete", "Patch", "Put", "Post" ],
      "UpstreamHeaderTransform": {
        "Referrer": "{RemoteIpAddress}",
        "X-Forwarded-For": "{RemoteIpAddress}"
      },
      "SwaggerKey": "crud"
    },
    {
      "__ServiceName": "Web Client",
      "DownstreamPathTemplate": "/{catchAll}",
      "DownstreamScheme": "ws",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 5001
        }
      ],
      "UpstreamPathTemplate": "/{catchAll}",
      "UpstreamHttpMethod": [ "GET", "POST", "PUT", "DELETE", "OPTIONS" ],
      "UpstreamHeaderTransform": {
        "Referrer": "{RemoteIpAddress}",
        "X-Forwarded-For": "{RemoteIpAddress}"
      }
    },
    {
      "__ServiceName": "Event Bus",
      "DownstreamPathTemplate": "/{everything}",
      "DownstreamScheme": "ws",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 5002
        }
      ],
      "UpstreamPathTemplate": "/event/{everything}",
      "UpstreamHttpMethod": [ "Get", "Delete", "Patch", "Put", "Post" ],
      "UpstreamHeaderTransform": {
        "Referrer": "{RemoteIpAddress}",
        "X-Forwarded-For": "{RemoteIpAddress}"
      }
    }
  ],
  "SwaggerEndPoints": [
    {
      "Key": "crud",
      "Config": [
        {
          "Name": "Rest API Server",
          "Version": "v1",
          "Url": "http://localhost:5000/swagger/v1/swagger.json"
        }
      ]
    }
  ]
}
```

**Reply by OnSive** [reply]
@raman-m could you please take a look at [my question](https://github.com/ThreeMammals/Ocelot/discussions/1748#discussioncomment-7328588)?

**Reply by raman-m** [reply]
> ```json
> {
>   "GlobalConfiguration": {
>     "BaseUrl": "http://localhost:80",
>     "UpstreamHeaderTransform": {
>       "Referrer": "{RemoteIpAddress}",
>       "X-Forwarded-For": "{RemoteIpAddress}"
>     }
>   },
> }
> ```
> 

**Important note!** Headers Transforms feature works for Routes only. **You cannot define in Global section!**
So, You are able to define Headers Transforms locally in route def only.

We have pending open PR #1659 and open issue #1658 which will add the ability to define global headers transformation.

-----
> The issue with the remote IP address appears when sending requests from the remote machine.

What is your remote machine? How do you host this machine? Is it visible my public IP address?
Maybe you host this remote machine in Docker container?

-----------
> When I tried using the Headers Transformation (about what you wrote in the other answer) feature **I didn't event get the headers in the ASP.NET backed**.
> That's why **I tried writing a delegate handler as no other option seemed to work** and I get still the same not working result.

Try to debug this method to understand the `{RemoteIpAddress}` placeholder: [Ocelot.Infrastructure.Placeholders.GetRemoteIpAddress()](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Infrastructure/Placeholders.cs#L82)
And let me know your debug & testing results plz!

**Reply by OnSive** [reply]
@raman-m thanks for your help. I found the problem which is rather embarising:

I'm using a blazor server-side server and a ASP.NET cor backend server. When I'm sending a request from the Blazor application to the backend then it can only be localhost as the request is **not** send from the client but instead from the blazor server-side server.

If you are interested I'm going to update you on my solution but as of now I don't have one. [SO question](https://stackoverflow.com/questions/77396725/forward-blazor-server-client-ip-address-to-backend)

**Reply by raman-m** [reply]
Sorry, I **am not** interested in this discussion anymore. You have got all answers from my side!
Good luck in development using Blazor!

**Comment by raman-m** [comment]
> ## Additional question
> **What would be a feasible way to forward the remote ip address?**
> as the:
> ```json
> "UpstreamHeaderTransform": {
>   "X-Forwarded-For": "{RemoteIpAddress}"
> }
> ```
> Configuration does not seem to do anything. (at least not in the ASP.NET backend)
> 

I try to answer...
You can do that with [Headers Transformation](https://ocelot.readthedocs.io/en/latest/features/headerstransformation.html) feature.
Especially this recipe: [X-Forwarded-For](https://ocelot.readthedocs.io/en/latest/features/headerstransformation.html#x-forwarded-for)
This feature should extract remote address as placeholder and forward as a header to your downstreams. It should work!
But, again, remember, if you've deployed all apps to local environment (development PC) then you will see `localhost`, or `192.168.x.x` or even `::1` because you are hosted app in local environment.

Also, I recommend you to look into this PR #1221 which has remote address helper extensions, so soon, we will have them in develop.
Maybe you will be inspired by this PR 😉
